streams/src/main/java/org/apache/kafka/streams/processor/internals/StoreChangelogReader.java

```diff
@@ -38,346 +39,682 @@
+                " active tasks for restoring");
+        }
 
-                log.debug("Calling restorer for partition {}", partition);
-                restorer.restoreStarted();
-            } else {
-                log.trace("Did not find checkpoint from changelog {} for store {}, rewinding to beginning.", partition, restorer.storeName());
+        final Set<TopicPartition> changelogsWithLimitOffsets = changelogs.entrySet().stream()
+            .filter(entry -> entry.getValue().stateManager.taskType() == AbstractTask.TaskType.STANDBY &&
+                entry.getValue().stateManager.changelogAsSource(entry.getKey()))
+            .map(Map.Entry::getKey).collect(Collectors.toSet());
+
+        updateLimitOffsetsForStandbyChangelogs(committedOffsetForChangelogs(changelogsWithLimitOffsets));
+    }
+
+    private void updateLimitOffsetsForStandbyChangelogs(final Map<TopicPartition, Long> committedOffsets) {
+        for (final ChangelogMetadata metadata : changelogs.values()) {
+            if (metadata.stateManager.taskType() == AbstractTask.TaskType.STANDBY &&
+                metadata.stateManager.changelogAsSource(metadata.changelogPartition) &&
+                committedOffsets.containsKey(metadata.changelogPartition)) {
+
+                final Long newLimit = committedOffsets.get(metadata.changelogPartition);
+                final Long previousLimit = metadata.restoreLimitOffset;
+
+                if (previousLimit != null && previousLimit > newLimit) {
+                    throw new IllegalStateException("Offset limit should monotonically increase, but was reduced for partition " +
+                        metadata.changelogPartition + ". New limit: " + newLimit + ". Previous limit: " + previousLimit);
+                }
+
+                metadata.restoreLimitOffset = newLimit;
+            }
+        }
+    }
+
+    private void initializeChangelogs(final Set<ChangelogMetadata> newPartitionsToRestore) {
+        if (newPartitionsToRestore.isEmpty())
+            return;
 
-                restoreConsumer.seekToBeginning(Collections.singletonList(partition));
-                needsPositionUpdate.add(restorer);
+        // for active changelogs, we need to find their end offset before transit to restoring
+        // if the changelog is on source topic, then its end offset should be the minimum of
+        // its committed offset and its end offset; for standby tasks that use source topics
+        // as changelogs, we want to initialize their limit offsets as committed offsets as well
+        final Set<TopicPartition> newPartitionsToFindEndOffset = new HashSet<>();
+        final Set<TopicPartition> newPartitionsToFindCommittedOffset = new HashSet<>();
+
+        for (final ChangelogMetadata metadata : newPartitionsToRestore) {
+            final TopicPartition partition = metadata.changelogPartition;
+
+            if (metadata.stateManager.taskType() == AbstractTask.TaskType.ACTIVE) {
```

Comment:
I originally put this since in a later PR where I consolidate global tasks into this then for TaskType.GLOBAL the logic would be different, but I guess I will just follow your suggestion where with a TODO so that later maybe we can have a better idea.